Cpp compiling error: expected primary-expression before 'char'.    _Generic macros problem

**Describe the bug**

Compile error if use call  LOG_DBG(...)  from cpp code.

Error example:
error: expected primary-expression before 'char'.  

**To Reproduce**
Simple try compile cpp file with code:
// Type your code here, or load an example.

// C program to illustrate macro function.
#include <stdio.h>
#define geeks(T) _Generic( (T), char: 1, int: 2, long: 3, default: 0)
int main(void)
{
    // char returns ASCII value which is int type. 
    printf("%d\n", geeks('A')); 
   
    // Here A is a string.
    printf("%d",geeks("A"));
      
    return 0;
} 

You can test it on different compilers her. I found that clang from 3.5 cpp support  _Generic - but result different if compile c file.
https://godbolt.org/z/jf1o41co5

**Expected behavior**


**Impact**

**Logs and console output**
../zephyr/include/sys/cbprintf_internal.h:75:4: error: expected primary-expression before 'char'
   75 |    char * : 1, \
      |    ^~~~
../zephyr/include/sys/util_internal.h:67:26: note: in definition of macro '__DEBRACKET'
   67 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
../zephyr/include/sys/util_internal.h:59:2: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   59 |  __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |  ^~~~~~~~~~~~~~~~~~~~
../zephyr/include/sys/util_internal.h:56:2: note: in expansion of macro '__COND_CODE'
   56 |  __COND_CODE(_ZZZZ##_flag, _if_0_code, _else_code)
      |  ^~~~~~~~~~~
../zephyr/include/sys/util_macro.h:133:2: note: in expansion of macro 'Z_COND_CODE_0'
  133 |  Z_COND_CODE_0(_flag, _if_0_code, _else_code)
      |  ^~~~~~~~~~~~~
../zephyr/include/sys/cbprintf_internal.h:110:2: note: in expansion of macro 'COND_CODE_0'
  110 |  COND_CODE_0(NUM_VA_ARGS_LESS_1(__VA_ARGS__), \
      |  ^~~~~~~~~~~
../zephyr/include/sys/util_loops.h:1075:2: note: in expansion of macro 'Z_CBPRINTF_IS_PCHAR'
 1075 |  fixed_arg0(x)
      |  ^~~~~~~~~~
../zephyr/include/sys/util_loops.h:28:2: note: in expansion of macro 'Z_FOR_EACH_EXEC'
   28 |  z_call(0, x, fixed_arg0, fixed_arg1)
      |  ^~~~~~
../zephyr/include/sys/util_loops.h:31:2: note: in expansion of macro 'Z_FOR_LOOP_1'
   31 |  Z_FOR_LOOP_1(z_call, sep, fixed_arg0, fixed_arg1, ##__VA_ARGS__) \
      |  ^~~~~~~~~~~~
../zephyr/include/sys/util_loops.h:36:2: note: in expansion of macro 'Z_FOR_LOOP_2'
   36 |  Z_FOR_LOOP_2(z_call, sep, fixed_arg0, fixed_arg1, ##__VA_ARGS__) \
      |  ^~~~~~~~~~~~
../zephyr/include/sys/util_loops.h:23:53: note: in expansion of macro 'Z_FOR_LOOP_3'
   23 |     _57, _58, _59, _60, _61, _62, _63, _64, N, ...) N
      |                                                     ^
../zephyr/include/sys/util_loops.h:1078:2: note: in expansion of macro 'Z_FOR_EACH_ENGINE'
 1078 |  Z_FOR_EACH_ENGINE(Z_FOR_EACH_EXEC, sep, F, _, __VA_ARGS__)
      |  ^~~~~~~~~~~~~~~~~
../zephyr/include/sys/util_macro.h:371:2: note: in expansion of macro 'Z_FOR_EACH'
  371 |  Z_FOR_EACH(F, sep, REVERSE_ARGS(__VA_ARGS__))
      |  ^~~~~~~~~~
../zephyr/include/sys/cbprintf_internal.h:95:3: note: in expansion of macro 'FOR_EACH'
   95 |  (FOR_EACH(Z_CBPRINTF_IS_PCHAR, (+),  __VA_ARGS__))
      |   ^~~~~~~~
../zephyr/include/sys/cbprintf_internal.h:112:7: note: in expansion of macro 'Z_CBPRINTF_HAS_PCHAR_ARGS'
  112 |    (((Z_CBPRINTF_HAS_PCHAR_ARGS(__VA_ARGS__) - skip) > 0)))
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~
../zephyr/include/sys/cbprintf.h:92:2: note: in expansion of macro 'Z_CBPRINTF_MUST_RUNTIME_PACKAGE'
   92 |  Z_CBPRINTF_MUST_RUNTIME_PACKAGE(skip, __VA_ARGS__)
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../zephyr/include/logging/log_msg2.h:400:6: note: in expansion of macro 'CBPRINTF_MUST_RUNTIME_PACKAGE'
  400 |  if (CBPRINTF_MUST_RUNTIME_PACKAGE(_cstr_cnt, __VA_ARGS__)) { \
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../zephyr/include/logging/log_msg2.h:429:3: note: in expansion of macro 'Z_LOG_MSG2_CREATE2'
  429 |   Z_LOG_MSG2_CREATE2(_try_0cpy, _mode, 1, _domain_id, _source, \
      |   ^~~~~~~~~~~~~~~~~~
../zephyr/include/logging/log_core.h:315:3: note: in expansion of macro 'Z_LOG_MSG2_CREATE'
  315 |   Z_LOG_MSG2_CREATE(!IS_ENABLED(CONFIG_USERSPACE), _mode, \
      |   ^~~~~~~~~~~~~~~~~
../zephyr/include/logging/log_core.h:330:2: note: in expansion of macro 'Z_LOG2'
  330 |  Z_LOG2(_level, \
      |  ^~~~~~
../zephyr/include/logging/log.h:71:25: note: in expansion of macro 'Z_LOG'
   71 | #define LOG_DBG(...)    Z_LOG(LOG_LEVEL_DBG, __VA_ARGS__)
      |                         ^~~~~
../modules/libcpp/zephyr/global_allocator_overloads.cpp:13:5: note: in expansion of macro 'LOG_DBG'
   13 |     LOG_DBG("size = %u -> %p", (unsigned)size, ptr);
      |     ^~~~~~~

**Environment (please complete the following information):**
 - OS:  Windows
 - Toolchain :gcc version 10.2.1 20201103 (release) (GNU Arm Embedded Toolchain 10-2020-q4-major)
 - Commit SHA : SHA-1: a379196b48ede6c9e1b308c5ed92e1ce20bec94e
 


**Additional context**
-

--- a/include/logging/log_core.h
+++ b/include/logging/log_core.h
@@ -107,13 +107,7 @@ extern "C" {
  *  @param _addr Address of the element.
  */
 #define LOG_CONST_ID_GET(_addr) \
-	Z_LOG_EVAL(\
-	  CONFIG_LOG,\
-	  (__log_level ? \
-	  log_const_source_id((const struct log_source_const_data *)_addr) : \
-	  0),\
-	  (0)\
-	)
+	COND_CODE_1(CONFIG_LOG, ((__log_level ? log_const_source_id(_addr) : 0)), (0))
 
 /**
  * @def LOG_CURRENT_MODULE_ID
@@ -134,12 +128,7 @@ extern "C" {
  *  @param _addr Address of the element.
  */
 #define LOG_DYNAMIC_ID_GET(_addr) \
-	Z_LOG_EVAL(\
-	  CONFIG_LOG,\
-	  (__log_level ? \
-	  log_dynamic_source_id((struct log_source_dynamic_data *)_addr) : 0),\
-	  (0)\
-	)
+	COND_CODE_1(CONFIG_LOG, ((__log_level ? log_dynamic_source_id(_addr) : 0)), (0))
 
 /* Set of defines that are set to 1 if function name prefix is enabled for given level. */
 #define Z_LOG_FUNC_PREFIX_0U 0
@@ -215,10 +204,10 @@ extern "C" {
 	} \
 } while (false)
 
-#define Z_LOG_INTERNAL(is_user_context, _level, _source, ...) do { \
+#define Z_LOG_INTERNAL(is_user_context, _level, _source, _dsource, ...) do { \
 	uint16_t src_id = \
 		IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-		LOG_DYNAMIC_ID_GET(_source) : LOG_CONST_ID_GET(_source);\
+		LOG_DYNAMIC_ID_GET(_dsource) : LOG_CONST_ID_GET(_source); \
 	struct log_msg_ids src_level = { \
 		.level = _level, \
 		.domain_id = CONFIG_LOG_DOMAIN_ID, \
@@ -303,7 +292,7 @@ static inline char z_log_minimal_level_to_char(int level)
 /*****************************************************************************/
 /****************** Macros for standard logging ******************************/
 /*****************************************************************************/
-#define Z_LOG2(_level, _source, ...) do { \
+#define Z_LOG2(_level, _source, _dsource, ...) do { \
 	if (!Z_LOG_CONST_LEVEL_CHECK(_level)) { \
 		break; \
 	} \
@@ -314,17 +303,20 @@ static inline char z_log_minimal_level_to_char(int level)
 	\
 	bool is_user_context = k_is_user_context(); \
 	uint32_t filters = IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-	     ((struct log_source_dynamic_data *)(void *)(_source))->filters : 0;\
+						(_dsource)->filters : 0;\
 	if (!LOG_CHECK_CTX_LVL_FILTER(is_user_context, _level, filters)) { \
 		break; \
 	} \
 	if (IS_ENABLED(CONFIG_LOG2)) { \
 		int _mode; \
+		void *_src = IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
+			(void *)_dsource : (void *)_source; \
 		Z_LOG_MSG2_CREATE(UTIL_NOT(IS_ENABLED(CONFIG_USERSPACE)), _mode, \
-				  CONFIG_LOG_DOMAIN_ID, _source, _level, NULL,\
+				  CONFIG_LOG_DOMAIN_ID, _src, _level, NULL,\
 				  0, __VA_ARGS__); \
 	} else { \
-		Z_LOG_INTERNAL(is_user_context,	_level, _source, __VA_ARGS__);\
+		Z_LOG_INTERNAL(is_user_context,	_level, \
+				_source, _dsource, __VA_ARGS__);\
 	} \
 	if (false) { \
 		/* Arguments checker present but never evaluated.*/ \
@@ -335,27 +327,27 @@ static inline char z_log_minimal_level_to_char(int level)
 } while (false)
 
 #define Z_LOG(_level, ...) \
-	Z_LOG2(_level, \
-	      IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-	      (void *)__log_current_dynamic_data : \
-	      (void *)__log_current_const_data, \
-	      __VA_ARGS__)
+	Z_LOG2(_level, __log_current_const_data, __log_current_dynamic_data, __VA_ARGS__)
 
 #define Z_LOG_INSTANCE(_level, _inst, ...) \
-	Z_LOG2(_level, Z_LOG_INST(_inst), __VA_ARGS__)
-
+	Z_LOG2(_level, \
+		COND_CODE_1(CONFIG_LOG_RUNTIME_FILTERING, (NULL), (Z_LOG_INST(_inst))), \
+		(struct log_source_dynamic_data *)COND_CODE_1( \
+						CONFIG_LOG_RUNTIME_FILTERING, \
+						(Z_LOG_INST(_inst)), (NULL)), \
+		__VA_ARGS__)
 
 /*****************************************************************************/
 /****************** Macros for hexdump logging *******************************/
 /*****************************************************************************/
-#define Z_LOG_HEXDUMP2(_level, _source, _data, _len, ...) do { \
+#define Z_LOG_HEXDUMP2(_level, _source, _dsource, _data, _len, ...) do { \
 	const char *_str = GET_ARG_N(1, __VA_ARGS__); \
 	if (!Z_LOG_CONST_LEVEL_CHECK(_level)) {	\
 		break; \
 	} \
 	bool is_user_context = k_is_user_context(); \
 	uint32_t filters = IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-	     ((struct log_source_dynamic_data *)(void *)(_source))->filters : 0;\
+						(_dsource)->filters : 0;\
 	\
 	if (IS_ENABLED(CONFIG_LOG_MINIMAL)) { \
 		Z_LOG_TO_PRINTK(_level, "%s", _str); \
@@ -368,8 +360,10 @@ static inline char z_log_minimal_level_to_char(int level)
 	} \
 	if (IS_ENABLED(CONFIG_LOG2)) { \
 		int mode; \
+		void *_src = IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
+			(void *)_dsource : (void *)_source; \
 		Z_LOG_MSG2_CREATE(UTIL_NOT(IS_ENABLED(CONFIG_USERSPACE)), mode, \
-				  CONFIG_LOG_DOMAIN_ID, _source, _level, \
+				  CONFIG_LOG_DOMAIN_ID, _src, _level, \
 				  _data, _len, \
 				COND_CODE_0(NUM_VA_ARGS_LESS_1(_, ##__VA_ARGS__), \
 					(), \
@@ -379,7 +373,7 @@ static inline char z_log_minimal_level_to_char(int level)
 	} \
 	uint16_t src_id = \
 		IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-		LOG_DYNAMIC_ID_GET(_source) : LOG_CONST_ID_GET(_source);\
+		LOG_DYNAMIC_ID_GET(_dsource) : LOG_CONST_ID_GET(_source);\
 	struct log_msg_ids src_level = { \
 		.level = _level, \
 		.domain_id = CONFIG_LOG_DOMAIN_ID, \
@@ -396,13 +390,18 @@ static inline char z_log_minimal_level_to_char(int level)
 } while (false)
 
 #define Z_LOG_HEXDUMP(_level, _data, _length, ...) \
-	Z_LOG_HEXDUMP2(_level, IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-		      (void *)__log_current_dynamic_data : \
-		      (void *)__log_current_const_data, \
+	Z_LOG_HEXDUMP2(_level, \
+		      __log_current_const_data, \
+		      __log_current_dynamic_data, \
 		      _data, _length, __VA_ARGS__)
 
 #define Z_LOG_HEXDUMP_INSTANCE(_level, _inst, _data, _length, _str) \
-	Z_LOG_HEXDUMP2(_level, Z_LOG_INST(_inst), _data, _length, _str)
+	Z_LOG_HEXDUMP2(_level, \
+		COND_CODE_1(CONFIG_LOG_RUNTIME_FILTERING, (NULL), (Z_LOG_INST(_inst))), \
+		(struct log_source_dynamic_data *)COND_CODE_1( \
+						CONFIG_LOG_RUNTIME_FILTERING, \
+						(Z_LOG_INST(_inst)), (NULL)), \
+		_data, _length, _str)
 
 /*****************************************************************************/
 /****************** Filtering macros *****************************************/
@@ -798,12 +797,11 @@ __syscall void z_log_hexdump_from_user(uint32_t src_level_val,
 /******************************************************************************/
 #define Z_LOG_VA(_level, _str, _valist, _argnum, _strdup_action)\
 	__LOG_VA(_level, \
-		 IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-		 (void *)__log_current_dynamic_data : \
-		 (void *)__log_current_const_data, \
+		 __log_current_const_data, \
+		 __log_current_dynamic_data, \
 		 _str, _valist, _argnum, _strdup_action)
 
-#define __LOG_VA(_level, _source, _str, _valist, _argnum, _strdup_action) do { \
+#define __LOG_VA(_level, _source, _dsource, _str, _valist, _argnum, _strdup_action) do { \
 	if (!Z_LOG_CONST_LEVEL_CHECK(_level)) { \
 		break; \
 	} \
@@ -814,7 +812,7 @@ __syscall void z_log_hexdump_from_user(uint32_t src_level_val,
 	\
 	bool is_user_context = k_is_user_context(); \
 	uint32_t filters = IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-	     ((struct log_source_dynamic_data *)(void *)(_source))->filters : 0;\
+						_dsource->filters : 0;\
 	if (!LOG_CHECK_CTX_LVL_FILTER(is_user_context, _level, filters)) { \
 		break; \
 	} \
@@ -825,7 +823,7 @@ __syscall void z_log_hexdump_from_user(uint32_t src_level_val,
 	} \
 	uint16_t _id = \
 		IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? \
-		LOG_DYNAMIC_ID_GET(_source) : LOG_CONST_ID_GET(_source);\
+		LOG_DYNAMIC_ID_GET(_dsource) : LOG_CONST_ID_GET(_source);\
 	struct log_msg_ids src_level = { \
 		.level = _level, \
 		.domain_id = CONFIG_LOG_DOMAIN_ID, \

--- a/include/logging/log_msg2.h
+++ b/include/logging/log_msg2.h
@@ -186,13 +186,12 @@ enum z_log_msg2_mode {
 /* Messages are aligned to alignment required by cbprintf package. */
 #define Z_LOG_MSG2_ALIGNMENT CBPRINTF_PACKAGE_ALIGNMENT
 
-
 #if CONFIG_LOG2_USE_VLA
 #define Z_LOG_MSG2_ON_STACK_ALLOC(ptr, len) \
 	long long _ll_buf[ceiling_fraction(len, sizeof(long long))]; \
 	long double _ld_buf[ceiling_fraction(len, sizeof(long double))]; \
 	ptr = (sizeof(long double) == Z_LOG_MSG2_ALIGNMENT) ? \
-			(void *)_ld_buf : (void *)_ll_buf; \
+			(struct log_msg2 *)_ld_buf : (struct log_msg2 *)_ll_buf; \
 	if (IS_ENABLED(CONFIG_LOG_TEST_CLEAR_MESSAGE_SPACE)) { \
 		/* During test fill with 0's to simplify message comparison */ \
 		memset(ptr, 0, len); \
@@ -215,15 +214,17 @@ enum z_log_msg2_mode {
 	long double _ld_buf128[128 / sizeof(long double)]; \
 	long double _ld_buf256[256 / sizeof(long double)]; \
 	if (sizeof(long double) == Z_LOG_MSG2_ALIGNMENT) { \
-		ptr = (len > 128) ? (void *)_ld_buf256 : \
-			((len > 64) ? (void *)_ld_buf128 : \
-			((len > 48) ? (void *)_ld_buf64 : \
-			((len > 32) ? (void *)_ld_buf48 : (void *)_ld_buf32)));\
+		ptr = (len > 128) ? (struct log_msg2 *)_ld_buf256 : \
+			((len > 64) ? (struct log_msg2 *)_ld_buf128 : \
+			((len > 48) ? (struct log_msg2 *)_ld_buf64 : \
+			((len > 32) ? (struct log_msg2 *)_ld_buf48 : \
+				      (struct log_msg2 *)_ld_buf32)));\
 	} else { \
-		ptr = (len > 128) ? (void *)_ll_buf256 : \
-			((len > 64) ? (void *)_ll_buf128 : \
-			((len > 48) ? (void *)_ll_buf64 : \
-			((len > 32) ? (void *)_ll_buf48 : (void *)_ll_buf32)));\
+		ptr = (len > 128) ? (struct log_msg2 *)_ll_buf256 : \
+			((len > 64) ? (struct log_msg2 *)_ll_buf128 : \
+			((len > 48) ? (struct log_msg2 *)_ll_buf64 : \
+			((len > 32) ? (struct log_msg2 *)_ll_buf48 : \
+				      (struct log_msg2 *)_ll_buf32)));\
 	} \
 	if (IS_ENABLED(CONFIG_LOG_TEST_CLEAR_MESSAGE_SPACE)) { \
 		/* During test fill with 0's to simplify message comparison */ \
@@ -254,7 +255,7 @@ enum z_log_msg2_mode {
 	} \
 	struct log_msg2_desc _desc = \
 		     Z_LOG_MSG_DESC_INITIALIZER(_domain_id, _level, \
-			   _plen, _dlen); \
+			   (uint32_t)_plen, _dlen); \
 	z_log_msg2_finalize(_msg, _source, _desc, _data); \
 } while (0)
 
@@ -276,7 +277,7 @@ do { \
 	} \
 	struct log_msg2_desc _desc = \
 		Z_LOG_MSG_DESC_INITIALIZER(_domain_id, _level, \
-					   _plen, _dlen); \
+					   (uint32_t)_plen, _dlen); \
 	LOG_MSG2_DBG("creating message on stack: package len: %d, data len: %d\n", \
 			_plen, (int)(_dlen)); \
 	z_log_msg2_static_create((void *)_source, _desc, _msg->data, _data); \
@@ -289,7 +290,7 @@ do { \
 	size_t _msg_wlen = Z_LOG_MSG2_ALIGNED_WLEN(_plen, 0); \
 	struct log_msg2 *_msg = z_log_msg2_alloc(_msg_wlen); \
 	struct log_msg2_desc _desc = \
-		Z_LOG_MSG_DESC_INITIALIZER(_domain_id, _level, _plen, 0); \
+		Z_LOG_MSG_DESC_INITIALIZER(_domain_id, _level, (uint32_t)_plen, 0); \
 	LOG_MSG2_DBG("creating message zero copy: package len: %d, msg: %p\n", \
 			_plen, _msg); \
 	if (_msg) { \
@@ -437,11 +438,9 @@ do { \
 
 #define Z_LOG_MSG2_CREATE(_try_0cpy, _mode,  _domain_id, _source,\
 			  _level, _data, _dlen, ...) \
-do { \
 	Z_LOG_MSG2_CREATE2(_try_0cpy, _mode, UTIL_CAT(Z_LOG_FUNC_PREFIX_, _level), \
 			   _domain_id, _source, _level, _data, _dlen, \
-			   Z_LOG_STR(_level, __VA_ARGS__));\
-} while (0)
+			   Z_LOG_STR(_level, __VA_ARGS__))
 
 #define Z_TRACING_LOG_HDR_INIT(name, id) \
 	struct log_msg2_trace name = { \

--- a/include/sys/cbprintf.h
+++ b/include/sys/cbprintf.h
@@ -26,9 +26,9 @@ extern "C" {
  * - Clang 3.0 https://releases.llvm.org/3.0/docs/ClangReleaseNotes.html
  */
 #ifndef Z_C_GENERIC
-#if ((__STDC_VERSION__ >= 201112L) || \
+#if !defined(__cplusplus) && (((__STDC_VERSION__ >= 201112L) || \
 	((__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__) >= 40900) || \
-	((__clang_major__ * 10000 + __clang_minor__ * 100 + __clang_patchlevel__) >= 30000))
+	((__clang_major__ * 10000 + __clang_minor__ * 100 + __clang_patchlevel__) >= 30000)))
 #define Z_C_GENERIC 1
 #else
 #define Z_C_GENERIC 0

